src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,958 @@
+                std::mem::replace(&mut remap_upper, frontiers[0].frontier().to_owned());
+            while let Some(update) = pending_remap.peek_mut() {
+                if !remap_upper.less_equal(&update.0 .0) {
+                    let Reverse((into, from, diff)) = PeekMut::pop(update);
+                    remap_trace.push((from, into, diff));
+                } else {
+                    break;
+                }
+            }
+
+            // STEP 3. Receive new source updates
+            //         The `events` input describes arbitrary progress and data over `FromTime`,
+            //         which must be translated to `IntoTime`. Each `FromTime` can be found as the
+            //         first `IntoTime` associated with a `[FromTime]` that is not less or equal to
+            //         the input `FromTime`. Received events that are not yet associated to an
+            //         `IntoTime` are collected, and formed into a "chain batch": a sequence of
+            //         chains that results from sorting the updates by `FromTime`, and then
+            //         segmenting the sequence at elements where the partial order on `FromTime` is
+            //         violated.
+            while let Some(event) = puller.pull() {
+                match event.as_mut() {
+                    Event::Progress(changes) => {
+                        source_frontier.update_iter(changes.drain(..));
+                    }
+                    Event::Messages(_, data) => stash.append(data),
+                }
+            }
+            stash.sort_unstable_by(|(_, t1, _): &(D, FromTime, R), (_, t2, _)| t1.cmp(t2));
+            let mut new_source_updates = ChainBatch::from_iter(stash.drain(..));
+
+            // STEP 4: Reclock new and deferred updates
+            //         We are now ready to step through the remap bindings in time order and
+            //         perform the following actions:
+            //         4.1. Match `new_source_updates` against the entirety of bindings contained
+            //              in the trace.
+            //         4.2. Match `deferred_source_updates` against the bindings that were just
+            //              added in the trace.
+            //         4.3. Reclock `source_frontier` to calculate the new since frontier of the
+            //              remap trace.
+            //
+            //         The steps above only make sense to perform if there are any times for which
+            //         we can correctly accumulate the remap trace, which is what we check here.
+            if remap_since.iter().all(|t| !remap_upper.less_equal(t)) {
+                let mut cur_time = IntoTime::minimum();
+                cur_time.advance_by(remap_since.borrow());
+                let mut cur_binding = MutableAntichain::new();
+
+                let mut remap = remap_trace.iter().peekable();
+                let mut reclocked_source_frontier = remap_upper.clone();
+                loop {
```

Comment:
The `loop {` constructs make it hard to read. If they are really necessary, could you add some comments about the conditions under which we leave the loop? In particular, it looks like we are morally just iterating through `remap`, if I look for the `break` statement.